suite/tests/client-interface/drbbdup-empty-bb-test.dll.c

```diff
@@ -0,0 +1,300 @@
+        case2_analysis_called = true;
+        break;
+    default: CHECK(false, "invalid encoding");
+    }
+
+    return DR_EMIT_DEFAULT;
+}
+
+static void
+destroy_analysis(void *drcontext, uintptr_t encoding, void *user_data,
+                 void *orig_analysis_data, void *analysis_data)
+{
+    CHECK(orig_analysis_data == ORIG_ANALYSIS_VAL, "orig analysis data does not match");
+
+    switch (encoding) {
+    case 0: CHECK(false, "should not be called because analysis data is NULL"); break;
+    case 1:
+        CHECK(analysis_data == ANALYSIS_VAL_1, "invalid encoding for case 1");
+        case1_analysis_destroy_called = true;
+        break;
+    case 2:
+        CHECK(analysis_data == ANALYSIS_VAL_2, "invalid encoding for case 2");
+        case2_analysis_destroy_called = true;
+        break;
+    default: CHECK(false, "invalid encoding");
+    }
+}
+
+static void
+update_encoding()
+{
+    if (encode_val != 0)
+        encode_val--;
+}
+
+static void
+insert_encode(void *drcontext, void *tag, instrlist_t *bb, instr_t *where,
+              void *user_data, void *orig_analysis_data)
+{
+    CHECK(orig_analysis_data == ORIG_ANALYSIS_VAL, "orig analysis data does not match");
+
+    if (!is_cur_empty)
+        dr_insert_clean_call(drcontext, bb, where, update_encoding, false, 0);
+}
+
+static void
+print_case(uintptr_t case_val)
+{
+    dr_fprintf(STDERR, "case %u\n", case_val);
+}
+
+static dr_emit_flags_t
+instrument_instr(void *drcontext, void *tag, instrlist_t *bb, instr_t *instr,
+                 instr_t *where, bool for_trace, bool translating, uintptr_t encoding,
+                 void *user_data, void *orig_analysis_data, void *analysis_data)
+{
+    bool is_first, is_first_nonlabel, is_last;
+    drbbdup_status_t res;
+
+    CHECK(!is_cur_empty, "should not be called for empty basic block");
+
+    CHECK(orig_analysis_data == ORIG_ANALYSIS_VAL, "orig analysis data does not match");
+
+    switch (encoding) {
+    case 0:
+        CHECK(analysis_data == NULL, "case analysis does not match for default case");
+        break;
+    case 1:
+        CHECK(analysis_data == ANALYSIS_VAL_1, "case analysis does not match for case 1");
+        break;
+    case 2:
+        CHECK(analysis_data == ANALYSIS_VAL_2, "case analysis does not match for case 2");
+        break;
+    default: CHECK(false, "invalid encoding");
+    }
+
+    res = drbbdup_is_first_instr(drcontext, instr, &is_first);
+    CHECK(res == DRBBDUP_SUCCESS, "failed to check whether instr is start");
+    CHECK(is_first, "must be first");
+
+    res = drbbdup_is_first_nonlabel_instr(drcontext, instr, &is_first_nonlabel);
+    CHECK(res == DRBBDUP_SUCCESS, "failed to check whether instr is first non-label");
+    CHECK(is_first_nonlabel, "must be first non-label");
+
+    res = drbbdup_is_last_instr(drcontext, instr, &is_last);
+    CHECK(res == DRBBDUP_SUCCESS, "failed to check whether instr is last");
+    CHECK(is_last, "must be last");
```

Comment:
I don't understand: this seems to assume that every block has either zero instrs or a single instr?  That doesn't seem right.

Reply:
OK now I understand: it relies on -max_bb_instrs 0.  Please add another comment here.

Reply:
Done.